tracer/src/Datadog.Trace/AppSec/Waf/Waf.cs

```diff
@@ -59,75 +55,94 @@ public string Version
-            }
-            else
-            {
-                initalizationResult = WafConfigurator.Configure(rulesFile, wafNative, encoder, obfuscationParameterKeyRegex, obfuscationParameterValueRegex);
-            }
-
-            return new Waf(initalizationResult, wafNative, encoder);
+            return InitializationResult.From(libraryHandle, rulesJson, rulesFile, obfuscationParameterKeyRegex, obfuscationParameterValueRegex);
         }
 
-        public IContext CreateContext()
+        /// <summary>
+        /// Requires a non disposed waf handle
+        /// </summary>
+        /// <returns>Context object to perform matching using the provided WAF instance</returns>
+        /// <exception cref="Exception">Exception</exception>
+        public IContext? CreateContext()
         {
-            var contextHandle = wafNative.InitContext(ruleHandle.Value);
+            if (Disposed)
+            {
+                Log.Warning("Context can't be created as waf instance has been disposed.");
+                return null;
+            }
+
+            var contextHandle = wafNative.InitContext(wafHandle);
 
             if (contextHandle == IntPtr.Zero)
             {
                 Log.Error(InitContextError);
                 throw new Exception(InitContextError);
             }
 
-            return new Context(contextHandle, wafNative, encoder);
+            return new Context(contextHandle, this);
         }
 
-        public void Dispose()
+        // Requires a non disposed waf handle
+        public bool UpdateRulesData(IEnumerable<RuleData> res)
         {
-            Dispose(true);
-            GC.SuppressFinalize(this);
-        }
+            if (Disposed)
+            {
+                Log.Warning("Waf instance has been disposed when trying to update rules data");
+                return false;
+            }
 
-        public bool UpdateRules(IEnumerable<RuleData> res)
-        {
-            if (res == null || res.Count() == 0)
+            if (res.Count() == 0)
             {
                 return false;
             }
 
             var finalRuleDatas = MergeRuleData(res);
             using var encoded = encoder.Encode(finalRuleDatas, new List<Obj>(), false);
-            var ret = wafNative.UpdateRuleData(ruleHandle.Value, encoded.RawPtr);
+            var ret = wafNative.UpdateRuleData(wafHandle, encoded.RawPtr);
             Log.Information("{number} rules have been updated and waf status is {status}", finalRuleDatas.Count, ret);
             return ret == DDWAF_RET_CODE.DDWAF_OK;
         }
 
+        /// <summary>
+        /// Requires a non disposed waf handle
+        /// </summary>
+        /// <param name="ruleStatus">whether rules have been toggled</param>
+        /// <returns>whether or not rules were toggled</returns>
         public bool ToggleRules(IDictionary<string, bool> ruleStatus)
         {
-            if (ruleStatus == null || ruleStatus.Count == 0)
+            if (Disposed)
```

Comment:
I think protection is a good idea, but we should try to use a read / write lock to attempt to ensure that it doesn't happen. If I've understood the PR correctly, this is the protection, it should be a back-up protection that we hope never happens.

Reply:
yes done [here](https://github.com/DataDog/dd-trace-dotnet/blob/6f27d605dc087825f9e2c8b7a585f5ef8cdb0337/tracer/src/Datadog.Trace/AppSec/Waf/Context.cs#L44)